internal/kube/client.go

```diff
@@ -0,0 +1,368 @@
+						var b strings.Builder
+						_, _ = fmt.Fprintf(&b, "%s:", status.Message)
+						for _, c := range status.Details.Causes {
+							_, _ = fmt.Fprintf(&b, " %s", c.Message)
+						}
+						reason = b.String()
+					}
+					return false, reason, nil
+				}
+				return false, err.Error(), nil
+			}
+			continue
+		}
+
+		_, err = ri.Patch(
+			ctx,
+			u.GetName(),
+			types.ApplyPatchType,
+			data,
+			metav1.PatchOptions{
+				DryRun:       []string{metav1.DryRunAll},
+				FieldManager: "datumctl-validate",
+			},
+		)
+		if err != nil {
+			// Treat API validation failures as ok=false with a descriptive message.
+			if statusErr, ok := err.(*apierrors.StatusError); ok {
+				status := statusErr.ErrStatus
+				reason := status.Message
+				if status.Details != nil && len(status.Details.Causes) > 0 {
+					var b strings.Builder
+					_, _ = fmt.Fprintf(&b, "%s:", status.Message)
+					for _, c := range status.Details.Causes {
+						_, _ = fmt.Fprintf(&b, " %s", c.Message)
+					}
+					reason = b.String()
+				}
+				return false, reason, nil
+			}
+			return false, err.Error(), nil
+		}
+	}
+
+	return true, fmt.Sprintf("validated %d object(s) (server-side dry-run)", docN), nil
+}
+
+// GetCRD returns the CRD by name formatted as yaml|json|describe (default: yaml).
+func (k *Kubectl) GetCRD(ctx context.Context, name, mode string) (string, error) {
+	if err := k.ensureClusterReachable(ctx); err != nil {
+		return "", err
+	}
+	crd, err := k.apix.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("get CRD %q: %w", name, err)
+	}
+	switch strings.ToLower(mode) {
+	case "", "yaml":
+		js, err := json.Marshal(crd)
+		if err != nil {
+			return "", err
+		}
+		yb, err := yaml.JSONToYAML(js)
+		if err != nil {
+			return "", err
+		}
+		return string(yb), nil
+	case "json":
+		js, err := json.MarshalIndent(crd, "", "  ")
+		if err != nil {
+			return "", err
+		}
+		return string(js), nil
+	case "describe":
+		return formatCRDDescribe(crd), nil
+	default:
+		return "", fmt.Errorf("unsupported mode %q (use yaml|json|describe)", mode)
+	}
+}
+
+// Preflight verifies API connectivity (no return payload).
+func (k *Kubectl) Preflight(ctx context.Context) error {
+	if err := k.ensureClusterReachable(ctx); err != nil {
+		return err
+	}
+	if _, err := k.disco.ServerVersion(); err != nil {
+		return fmt.Errorf("server version: %w", err)
+	}
+	return nil
+}
+
+
+// ensureClusterReachable initializes clients and verifies the API server is reachable.
+func (k *Kubectl) ensureClusterReachable(ctx context.Context) error {
+	if err := k.initClients(); err != nil {
+		return err
+	}
+	if _, err := k.disco.ServerVersion(); err != nil {
+		return fmt.Errorf("cannot reach Kubernetes API for the selected context: %w", err)
+	}
+	return nil
+}
+
+// initClients loads kubeconfig and constructs client-go clients on first use.
+func (k *Kubectl) initClients() error {
```

Comment:
This seems like something that should be done in a client creator (e.g. [NewUserContextualClient](https://github.com/datum-cloud/datumctl/blob/main/internal/client/user_context.go#L18)).